From 203d6991562928cdbbef08a39b21a6d02050e7bd Mon Sep 17 00:00:00 2001
From: Zhao Yan <yan.y.zhao@intel.com>
Date: Wed, 8 Jun 2016 16:01:39 +0800
Subject: [PATCH 4/5] fix the bug that qemu will crash when reboot windows
 guest frequently

in xen_ram_addr_from_mapcache() when translating host address ptr to
guest address, there may exist two matching entries in mapcache->entry of
the same paddr_index and size. so need to compare more fields to ensure
the correct entry is found.

Signed-off-by: Zhao Yan <yan.y.zhao@intel.com>
Signed-off-by: Bing Niu <bing.niu@intel.com>
diff --git a/xen-mapcache.c b/xen-mapcache.c
index 8f3a592..2245f45 100644
--- a/xen-mapcache.c
+++ b/xen-mapcache.c
@@ -337,7 +337,13 @@ ram_addr_t xen_ram_addr_from_mapcache(void *ptr)
     }
 
     entry = &mapcache->entry[paddr_index % mapcache->nr_buckets];
-    while (entry && (entry->paddr_index != paddr_index || entry->size != size)) {
+    while (entry && (entry->paddr_index != paddr_index || entry->size != size  ||
+            !(((unsigned long)ptr >= (unsigned long)entry->vaddr_base) &&
+              ((unsigned long) ptr < (unsigned long) entry->vaddr_base + (unsigned long)entry->size)) ||
+              !test_bits(((unsigned long) ptr - (unsigned long) entry->vaddr_base) >> XC_PAGE_SHIFT,
+              1,
+             entry->valid_mapping))) {
+
         entry = entry->next;
     }
     if (!entry) {
@@ -383,7 +389,12 @@ static void xen_invalidate_map_cache_entry_unlocked(uint8_t *buffer)
     }
 
     entry = &mapcache->entry[paddr_index % mapcache->nr_buckets];
-    while (entry && (entry->paddr_index != paddr_index || entry->size != size)) {
+    while (entry && (entry->paddr_index != paddr_index || entry->size != size ||
+        !(((unsigned long)buffer >= (unsigned long)entry->vaddr_base)
+        && ((unsigned long) buffer < (unsigned long) entry->vaddr_base + (unsigned long)entry->size)) ||
+        !test_bits(((unsigned long) buffer - (unsigned long) entry->vaddr_base) >> XC_PAGE_SHIFT,
+        1,
+        entry->valid_mapping))) {
         pentry = entry;
         entry = entry->next;
     }
@@ -431,23 +442,28 @@ void xen_invalidate_map_cache(void)
     for (i = 0; i < mapcache->nr_buckets; i++) {
         MapCacheEntry *entry = &mapcache->entry[i];
 
-        if (entry->vaddr_base == NULL) {
-            continue;
-        }
-        if (entry->lock > 0) {
-            continue;
-        }
-
-        if (munmap(entry->vaddr_base, entry->size) != 0) {
-            perror("unmap fails");
-            exit(-1);
+        while(entry){
+            if (entry->vaddr_base == NULL) {
+                entry = entry->next;
+                continue;
+            }
+            if (entry->lock > 0) {
+                entry = entry->next;
+                continue;
+            }
+
+            if (munmap(entry->vaddr_base, entry->size) != 0) {
+                perror("unmap fails");
+                exit(-1);
+            }
+
+            entry->paddr_index = 0;
+            entry->vaddr_base = NULL;
+            entry->size = 0;
+            g_free(entry->valid_mapping);
+            entry->valid_mapping = NULL;
+            entry = entry->next;
         }
-
-        entry->paddr_index = 0;
-        entry->vaddr_base = NULL;
-        entry->size = 0;
-        g_free(entry->valid_mapping);
-        entry->valid_mapping = NULL;
     }
 
     mapcache->last_entry = NULL;
